File: ios/sdk/cpp/KeypairWrapper.cpp
//
//  Elastos.SDK.Contact.cpp
//
//  Created by mengxk on 19/03/16.
//  Copyright © 2016 mengxk. All rights reserved.
//

#include <KeypairWrapper.hpp>

#include "Log.hpp"
#include <Elastos.SDK.Keypair.C/Elastos.Wallet.Utility.h>

namespace crosspl {
namespace native {
    
/***********************************************/
/***** static variables initialize *************/
/***********************************************/

/***********************************************/
/***** static function implement ***************/
/***********************************************/

int KeypairWrapper::GetSinglePublicKey(const std::span<uint8_t>* seed, std::stringstream* pubKey)
{
  pubKey->str("");

  auto ret = ::getSinglePublicKey(seed->data(), (int)seed->size());
  if(ret == nullptr) {
    return -1;
  }
  
  (*pubKey) << ret;
  freeBuf(ret);

  return 0;
}

int KeypairWrapper::GetSinglePrivateKey(const std::span<uint8_t>* seed, std::stringstream* privKey)
{
  privKey->str("");

  auto ret = ::getSinglePrivateKey(seed->data(), (int)seed->size());
  if(ret == nullptr) {
    return -1;
  }
  
  (*privKey) << ret;
  freeBuf(ret);

  return 0;
}

int KeypairWrapper::GenerateMnemonic(const char* language, const char* words, std::stringstream* mnem)
{
  mnem->str("");

  auto ret = ::generateMnemonic(language, words);
  if(ret == nullptr) {
    return -1;
  }
  
  (*mnem) << ret;
  freeBuf(ret);

  return 0;
}

int KeypairWrapper::GetSeedFromMnemonic(const char* mnemonic, const char* mnemonicPassword, std::vector<uint8_t>* seed)
{
  seed->clear();

  void* seedData = nullptr;
  int seedSize = ::getSeedFromMnemonic(&seedData, mnemonic, mnemonicPassword);
  if(seedData == nullptr || seedSize <= 0) {
    return -1;
  }

  *seed = std::vector<uint8_t>((uint8_t*)seedData, (uint8_t*)seedData + seedSize);

  freeBuf(seedData);

  return 0;
}

int KeypairWrapper::Sign(const char* privateKey, const std::span<uint8_t>* data, std::vector<uint8_t>* signedData)
{
  signedData->clear();

  void* keypairSignedData = nullptr;
  int keypairSignedSize = ::sign(privateKey, data->data(), data->size(), &keypairSignedData);
  if(keypairSignedData == nullptr || keypairSignedSize <= 0) {
    return -1;
  }

  *signedData = std::vector<uint8_t>((uint8_t*)keypairSignedData, (uint8_t*)keypairSignedData + keypairSignedSize);

  freeBuf(keypairSignedData);

  return 0;
}


/***********************************************/
/***** class public function implement  ********/
/***********************************************/

/***********************************************/
/***** class protected function implement  *****/
/***********************************************/


/***********************************************/
/***** class private function implement  *******/
/***********************************************/


} //namespace native
} //namespace crosspl

File: ios/sdk/swift/internal/KeypairWrapper.swift
import CrossPL

/* @CrossClass */
public class KeypairWrapper: CrossBase {

  /* @CrossNativeInterface */
  public static func GetSinglePublicKey(seed: Data, pubKey: inout String) -> Int {
    var nsPubKey = pubKey as NSString?
    let ret = crosspl_Proxy_KeypairWrapper_GetSinglePublicKey(seed, &nsPubKey)
    if ret < 0 {
      return Int(ret)
    }
    pubKey = nsPubKey! as String
    
    return 0
  }
  
  /* @CrossNativeInterface */
  public static func GetSinglePrivateKey(seed: Data, privKey: inout String) -> Int {
    var nsPrivKey = privKey as NSString?
    let ret = crosspl_Proxy_KeypairWrapper_GetSinglePrivateKey(seed, &nsPrivKey)
    if ret < 0 {
      return Int(ret)
    }
    privKey = nsPrivKey! as String
    
    return 0
  }
  
  /* @CrossNativeInterface */
  public static func GenerateMnemonic(language: String, words: String?, mnem: inout String?) -> Int {
    var nsMnem = mnem as NSString?
    let ret = crosspl_Proxy_KeypairWrapper_GenerateMnemonic(language, words, &nsMnem)
    if ret < 0 {
      return Int(ret)
    }
    mnem = nsMnem! as String
    
    return 0  }
  
  /* @CrossNativeInterface */
  public static func GetSeedFromMnemonic(mnemonic: String,
                                         mnemonicPassword: String,
                                         seed: inout Data) -> Int {
    var nsseed = seed as NSData?
    let ret = crosspl_Proxy_KeypairWrapper_GetSeedFromMnemonic(mnemonic, mnemonicPassword, &nsseed)
    if ret < 0 {
      return Int(ret)
    }
    seed = nsseed! as Data
    
    return Int(ret)
  }
  
  /* @CrossNativeInterface */
  public static func Sign(privateKey: String?, data: Data, signedData: inout Data) -> Int {
    var nsSignedData = signedData as NSData?
    let ret = crosspl_Proxy_KeypairWrapper_Sign(privateKey, data, &nsSignedData)
    if ret < 0 {
      return Int(ret)
    }
    signedData = nsSignedData! as Data
    
    return Int(ret)
  }
  
  init() {
    super.init(className: String(describing: ContactDebug.self))
  }
} // class Factory
